doc/sphinx/proof-engine/vernacular-commands.rst

```diff
@@ -185,194 +166,155 @@ Newly opened modules and sections inherit the current settings.
+     :n:`@one_term` in which holes of the pattern are indicated by `_` or :n:`?@ident`.
+     If the same :n:`?@ident` occurs more than once in the pattern, all occurrences must
+     match the same value.
 
-   .. cmdv:: Search {+ {? -}@term_pattern_string} inside {+ @qualid }
+   * :n:`@string` - If :n:`@string` is a substring of a valid identifier,
+     search for objects whose name contains :n:`@string`. If :n:`@string` is a notation
+     string associated with a :n:`@qualid`, that's equivalent to :cmd:`Search` :n:`@qualid`.
+     For example, specifying "+" or "_ + _", which are notations for "plus", are equivalent
+     to :cmd:`Search` :n:`plus`.
 
-      This restricts the search to constructions defined in the modules
-      named by the given :n:`qualid` sequence.
+   * :n:`% @scope` - limits the search to the scope bound to
+     the delimiting key :n:`@scope`, such as, for example, :n:`%nat`.
+     This clause may be used only if :n:`@string` contains a notation string.
+     (see Section :ref:`LocalInterpretationRulesForNotations`)
 
-   .. cmdv:: Search {+ {? -}@term_pattern_string} outside {+ @qualid }
+   If you specify multiple :n:`@search_item`\s, all the conditions must be satisfied
+   for the object to be displayed.  `{? - }` excludes objects that contain the :n:`@search_item`.
 
-      This restricts the search to constructions not defined in the modules
-      named by the given :n:`qualid` sequence.
+   Additional clauses:
 
-   .. cmdv:: @selector: Search {+ {? -}@term_pattern_string}
+   * :n:`inside {+ @qualid }` - limit the search to the specified modules
+   * :n:`outside {+ @qualid }` - exclude the specified modules from the search
 
-      This specifies the goal on which to search hypothesis (see
-      Section :ref:`invocation-of-tactics`).
-      By default the 1st goal is searched. This variant can
-      be combined with other variants presented here.
+   .. exn:: Module/section @qualid not found.
 
-      .. example::
+      There is no constant in the environment named :n:`@qualid`, where :n:`@qualid`
+      is in an `inside` or `outside` clause.
 
-         .. coqtop:: in
+   .. example:: :cmd:`Search` examples
 
-            Require Import ZArith.
+      .. coqtop:: in
 
-         .. coqtop:: all
+         Require Import ZArith.
 
-            Search Z.mul Z.add "distr".
+      .. coqtop:: all
 
-            Search "+"%Z "*"%Z "distr" -positive -Prop.
+         Search Z.mul Z.add "distr".
+         Search "+"%Z "*"%Z "distr" -Prop.
+         Search (?x * _ + ?x * _)%Z outside OmegaLemmas.
 
-            Search (?x * _ + ?x * _)%Z outside OmegaLemmas.
 
+.. cmd:: SearchHead @one_term {? {| inside | outside } {+ @qualid } }
 
-.. cmd:: SearchHead @term
+   Displays the name and type of all hypotheses of the
+   selected goal (if any) and theorems of the current context that have the
+   form :n:`{? forall {* @binder }, } {* P__i -> } C` where :n:`@one_term` matches `C`.
```

Comment:
My impression is that `@one_term` is a pattern and that it's pattern-matched to C.  Would it help to say "where the pattern `@one_term` ..."?

`@one_term {* _ }`, if it is a precise statement, introduces the idea of substitution into the pattern which then may need a little explanation here.  Or perhaps I'm confused?

Reply:
You are correct that `one_term` is a pattern, but what you are missing (in this explanation at least) is that with `SearchHead` it may be matched against a "head" of C, i.e., a non-fully applied term. What I meant by `@one_term (* _ }` is that what matches C is `one_term` applied to the right number of holes. See the examples I gave you previously and those which are in the doc.

Reply:
A concrete example (not necessarily runnable) that doesn't fit the wording is often the easiest for me to understand.

Are you referring to your example `SearchHead (_ = _ -> _ + _ = _ + _).` and that the pattern includes `->` whereas the wording suggests that C doesn't contain `->` (at least not at the top level)? Didn't see anything in the SearchHead examples in the doc that seem relevant.

By "head", I expect you mean starting from the root of the expression.  Don't see how "non-fully applied term" fits in here; IIUC your phrase refers to creating new function by providing some of its initial arguments similar to OCaml.

(FWIW GitHub doesn't make it so easy to find things in big conversations; some conversations are not loaded, then you have to unhide things before you search in the browser window.  Similar in Gmail--older message headers have to be expanded first.)

Reply:
> Are you referring to your example `SearchHead (_ = _ -> _ + _ = _ + _).` and that the pattern includes `->` whereas the wording suggests that C doesn't contain `->` (at least not at the top level)?

Not this one but the two in the doc are relevant:

```
SearchHead le.
SearchHead (@eq bool).
```

none of these would return any result when using `SearchPattern` instead of `SearchHead`.

>By "head", I expect you mean starting from the root of the expression.

No, I said "head of C".

Reply:
So for instance, C is `n <= m` and it wouldn't be matched by the pattern `le`, but it is matched by `le _ _` (same as `_ <= _`) and thus by `SeachHead le`. Similarly for `@eq bool` which matches any `_ = _` between bools.

Reply:
> >By "head", I expect you mean starting from the root of the expression.

> No, I said "head of C".

I was asking to confirm the definition of "head" and by "expression" I meant C.  "Head of C" just means the root of the expression tree for C?

> So for instance, C is `n <= m` and it wouldn't be matched by the pattern `le` ...

OK, I get the example, thanks.

I intended that "matches" in ```where :n:`@one_term` matches `C` ``` mean how Coq does matching for this command--a defined term, not meaning "equal".  If we use the unfamiliar `(@one_term t__1 ... t__m)` notation, where is that explained?  Perhaps patterns define functions (even for terms that don't look like functions) and this application just means substitution?  If we introduce the `t__1 ... t__m` there should be some explanation.

If other commands use the same type of matching, the description should be in its own section and hyperlinked.

Reply:
> > 
> 
> I was asking to confirm the definition of "head" and by "expression" I meant C. "Head of C" just means the root of the expression tree for C?

Yes.



> I intended that "matches" in `` where :n:`@one_term` matches `C` `` mean how Coq does matching for this command--a defined term, not meaning "equal". If we use the unfamiliar `(@one_term t__1 ... t__m)` notation, where is that explained? In either case it needs to be explained.

Sure. In fact, there is a missing section about unification in the "Language extensions" chapter, which would explain all that stuff. "match" would be defined there. It means unify two terms where some parts of the terms may be holes. "le" does not match "le _ _".

>  If we introduce the `t__1 ... t__m` there should be some explanation.

In fact, my preference would be to have `_ ... _` to express the additional holes.

Reply:
OK.  What stopgap wording can I do here for now?  I still lean towards calling it C since I don't have to explain the Tm or `_ ... _.  I could say "unifies by matching" and give your example as a note below.

I guess `le` matches a reference to that definition while `le _ _` matches a node with 2 subterms.  Are there other important cases worth mentioning without going too deep?

Reply:
> I could say "matches by unification".

No "matches by unification" would be exactly the same as "matches". You could say "the pattern matches a head of `C` (i.e., the pattern applied to sufficiently many holes matches `C`)."

>Are there other important cases worth mentioning without going too deep?

I don't think so.

Reply:
Not quite there yet.  I don't fully understand the description you suggest.

There's only one head for any expression, so it should be "the head", not "a head", right?  Nor do we define "head" anywhere in the document AFAIK.  But that should not be defined here.  Even so, if we left out "a/the head", I think readers would expect that the entire expression C has to be matched, not merely a subterm of C (cf. use of "subterms" in `Search`).

> Perhaps patterns define functions (even for terms that don't look like functions) and this application just means substitution?

It would really help if you would confirm or correct this guess.

If the parenthetical comment wasn't there, does that change the meaning?  If so, would you give a concrete example that would be incorrectly allowed without the comment?  Or is it just additional explanation for the user?

"sufficiently many holes" - I don't see why this is important.  If the pattern is considered a function, the *holes* are the parameters and the function is applied to *values* passed parameters.  I suppose you could partially apply a pattern function to get a new function, but that doesn't seem relevant.

Your thoughts?

Reply:
> There's only one head for any expression, so it should be "the head", not "a head", right?

Generally yes, the head of an expression is what you get at the beginning when you have decomposed your expression into applications. So in `x t_1 t_2 t_3`, the head is `x`. But in the case of `SearchHead`, the pattern can match the head of "C" *or* the head of "C" + the few first sub-terms to which it is applied in C *or* the entire "C". That's why I wrote "a head", as a generalization of the previous definition. But I agree that this is not very clear.

> not merely a subterm of C

That's not any subterm though.

>Perhaps patterns define functions (even for terms that don't look like functions) and this application just means substitution?

I'm sorry but I don't understand what this means.

>If the parenthetical comment wasn't there, does that change the meaning?

I'm not sure which one you are talking about. Are you taking about "(i.e., the pattern applied to sufficiently many holes matches C)". It does not change the meaning, it specifies what I meant "the pattern matches *a* head".

>"sufficiently many holes" - I don't see why this is important. If the pattern is considered a function, the holes are the parameters and the function is applied to values passed parameters. I suppose you could partially apply a pattern function to get a new function, but that doesn't seem relevant.

It seems to me that you are going into considerations that are way too complex.
Have you ever done some Prolog? If yes, it helps to understand unification, but if not let me explain this quickly.
Unification works like this: it tries to match two terms. These terms are built like normal Coq terms but may contain holes represented by existential variables (`?x`) or underscores (`_`). We can call such terms "patterns".
When matching, `le _ _` with `le a b` and the first `_` is replaced by `a` and the second `_` is replaced by `b`. On the other hand, when trying to match `le` with `le a b` or `le` with `le _ _`, we get a unification failure: the terms do not have the same form.

`Search` will match any sub-term. For instance, the pattern `a` will match the sub-term `a` of `le a b`.
`SearchPattern` will only match sub-terms that are *a* conclusion of the term being tested. For instance, `le _ _` will match the sub-term `le a b` in `forall a b, a = b -> le a b`. I say "a conclusion" because `_ = _ -> le _ _` will also work (it will match the sub-term `a = b -> le a b` which we can consider *a* conclusion but not *the* conclusion).
`SearchHead` will also match sub-terms that are *a* head of the conclusion of the sub-term being matched. For instance, `le` and `le _` will also work because applied to sufficiently many `_` they become patterns that match the conclusion `le a b`.

Reply:
> >Perhaps patterns define functions (even for terms that don't look like functions) and this application just means substitution?

> I'm sorry but I don't understand what this means.

Then what do you mean by "applied" in "`le` and `le _` will also work because applied to sufficiently many `_` they become patterns that match the conclusion `le a b`"?

> Have you ever done some Prolog?

If I did it was a long time ago.  I'm more familiar with unification in the context of resolution theorem proving, in which a pattern element with n variables could only match a function with n arguments--without any currying.

> Unification works like this: ... we get a unification failure: the terms do not have the same form.

Yup.

> It seems to me that you are going into considerations that are way too complex.

I want to say something that's both easy to understand but very accurate, so I need to understand it rather fully.

OK, "head" doesn't mean what I thought.

The additional holes only apply to the top node of the pattern, right?  For example `SearchHead (_ + _ = (Nat.add _)).` from your earlier examples doesn't match anything, while `SearchHead (_ + _ = (Nat.add _ _)).` does.

And (le _) would not match (le a b) under SearchPattern, right?

I don't see a way to make this clear with making the meaning of "a head" and "apply (a pattern)" explicit.  Could we a phrase like (roughly) "a prefix of the root node of C"??  Otherwise another sentence or two to explain.

Reply:
> > 
> 
> Then what do you mean by "applied" in "`le` and `le _` will also work because applied to sufficiently many `_` they become patterns that match the conclusion `le a b`"?

There's only one notion of application AFAIK and it is defined in https://coq.github.io/doc/master/refman/language/gallina-specification-language.html#applications

> The additional holes only apply to the top node of the pattern, right? For example `SearchHead (_ + _ = (Nat.add _)).` from your earlier examples doesn't match anything, while `SearchHead (_ + _ = (Nat.add _ _)).` does.

Correct.

>And (le _) would not match (le a b) under SearchPattern, right?

Correct.

>I don't see a way to make this clear with making the meaning of "a head" and "apply (a pattern)" explicit. Could we a phrase like (roughly) "a prefix of the root node of C"?? 

Sure, why not? It seems to me that whatever it contains, the explanation will need to rely on examples to illustrate anyway.

But please, let's not spend too much more time on this, it's getting tiring that this whole PR is blocked on such a minor thing.

Reply:
> But please, let's not spend too much more time on this, it's getting tiring that this whole PR is blocked on such a minor thing.

I agree.  It took a lot of back and forth to get to a common understanding--which I think was necessary and valuable.  Hopefully future discussions on other topics are more efficient.

How is this?

   Displays the name and type of all hypotheses of the
   selected goal (if any) and theorems of the current context that have the
   form :n:`{? forall {* @binder }, } {* P__i -> } C` where :n:`@one_term`
   matches a head of `C`.
   
   A *head* of `C` is the entire term `C` or a prefix of the top node of `C`.
   For example, a :n:`@one_term` of `(le _)` matches a `C` that is `(le (a+1) b)`.

Reply:
I think that my description using "a head" is too non-standard to include in the refman as is. Furthermore, your explanation of it is too imprecise. I don't know what "top node" means. And it does not talk about application. The example is not very natural because `le _` and `le` have the same behavior (when used with `SearchHead`) so the only natural thing is to give just `le`. Maybe `@eq bool` is a better example (cf. the examples contained in the manual).

Reply:
Look, I don't have any great suggestions but let's still settle in with a description, even if a bit heavy. Given that there might be changes to the `Search` infrastructure in 8.12, we will anyway have to rethink the whole doc.

Reply:
> There's only one notion of application AFAIK and it is defined in ...

I improved the wording there and added a mention of currying, which is barely mentioned in the documentation and may be unfamiliar to some users:

:n:`@term__fun @term` denotes applying the function :n:`@term__fun` to :token:`term`.

:n:`@term__fun {+ @term__i }` denotes applying :n:`@term__fun` to the arguments :n:`@term__i`.
By *currying*, this is equivalent to :n:`( … ( @term__fun @term__1 ) … ) @term__n`:
associativity is to the left.  (Any function of two arguments `f a b` can be converted to
the expression `(f a) b` in which `(f a)` represents a one argument function derived from `f`
by fixing the first argument to `a`.  Applying this repeatedly can convert the application
of any multi-argument function to a series of one argument functions.)

Reply:
> > Perhaps patterns define functions

You can apply a pattern to create a new, more complex pattern as shown below.  It's not the same as applying a function.  That's why I questioned the language "apply a pattern".  These are equivalent:

```
SearchHead (_ + _ = ((Nat.add _ _))).
SearchHead (_ + _ = ((Nat.add _) _)).  (* pattern "Nat.add _" applied to "_" *)
```

If a pattern match of this type is useful in proofs then it should be documented.  Also note that patterns in SearchHead are non-linear while those in `match` constructs are not.

Reply:
> By _currying_, this is equivalent

This is not equivalent by currying, the parentheses are implicit, but they are there if you print the term in full. I agree that this should be explained better as well, but please let's not put this too in this PR.

(Currying is the equivalence between fun x => fun y => and fun (x, y) =>)

> These are equivalent

These are not just equivalent. These are *exactly* the same. There is strictly no difference beyond writing explicitly the parentheses or not. This is standard term application. There is nothing specific to patterns here.

Reply:
> I think that my description using "a head" is too non-standard to include in the refman as is. Furthermore, your explanation of it is too imprecise. ...

How about this?  It is less precise. I define "prefix" implicitly, but the imprecision should be apparent to the reader, so at least it doesn't give a false impression of precision.  (Not 100% sure what you were referring to about @eq examples.)

```
   Displays the name and type of all hypotheses of the
   selected goal (if any) and theorems of the current context that have the
   form :n:`{? forall {* @binder }, } {* P__i -> } C` where :n:`@one_term`
   matches a prefix of `C`.  For example, a :n:`@one_term` of `(@eq _)`
   matches `@eq (and a b))`, which is a prefix of `C` when `C` is `(@eq (and a b) c)`.
```

Alternatively we could add several sentences defining various terms precisely.  I lean toward punting.

Reply:
Re: currying

OK, I'll just make it this slight wording improvement:

```
:n:`@term__fun @term` denotes applying the function :n:`@term__fun` to :token:`term`.

:n:`@term__fun {+ @term__i }` denotes applying
:n:`@term__fun` to the arguments :n:`@term__i`.  It is
equivalent to :n:`( … ( @term__fun @term__1 ) … ) @term__n`:
associativity is to the left.
```

> These are not just equivalent. These are *exactly* the same.

Well... Internally, they are the same.  From the users' perspective they are only equivalent--unless they print the term in full.

Reply:
This does look rather good to me, except that the example makes even less sense that the previous one. Putting the `@` in front of `eq` means that the first argument is the type of the equality. My goal was simply that the last argument in the pattern wouldn't be a `_` because it wouldn't be natural to put it there (since it wouldn't change anything).

Here's a modification of your proposal using completely abstract stuff:

```
   Displays the name and type of all hypotheses of the
   selected goal (if any) and theorems of the current context that have the
   form :n:`{? forall {* @binder }, } {* P__i -> } C` where :n:`@one_term`
   matches a prefix of `C`.  For example, a :n:`@one_term` of `f _ b`
   matches `f a b`, which is a prefix of `C` when `C` is `f a b c`.
```

Reply:
Then I'll use this wording, include the "application" wording tweak, move Search and squash.  It will take me a few minutes.